base/compiler/abstractinterpretation.jl

```diff
@@ -567,21 +565,30 @@ function abstract_iteration(interp::AbstractInterpreter, @nospecialize(itft), @n
     stateordonet === Bottom && return Any[Bottom]
     valtype = statetype = Bottom
     ret = Any[]
-    stateordonet = widenconst(stateordonet)
-    while !(Nothing <: stateordonet) && length(ret) < InferenceParams(interp).MAX_TUPLE_SPLAT
-        if !isa(stateordonet, DataType) || !(stateordonet <: Tuple) || isvatuple(stateordonet) || length(stateordonet.parameters) != 2
+    # Try to unroll the iteration up to MAX_TUPLE_SPLAT, which covers any finite
+    # length iterators, or interesting prefix
+    while true
+        stateordonet_widened = widenconst(stateordonet)
+        if Nothing <: stateordonet_widened || length(ret) >= InferenceParams(interp).MAX_TUPLE_SPLAT
+            break
+        end
+        if !isa(stateordonet_widened, DataType) || !(stateordonet_widened <: Tuple) || isvatuple(stateordonet_widened) || length(stateordonet_widened.parameters) != 2
             break
         end
-        if stateordonet.parameters[2] <: statetype
-            # infinite (or failing) iterator
+        nstatetype = getfield_tfunc(stateordonet, Const(2))
+        # If there's no new information in this statetype, don't bother continuing,
+        # the iterator won't be finite.
+        if nstatetype ⊑ statetype
             return Any[Bottom]
         end
-        valtype = stateordonet.parameters[1]
-        statetype = stateordonet.parameters[2]
+        valtype = getfield_tfunc(stateordonet, Const(1))
         push!(ret, valtype)
+        statetype = nstatetype
         stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], vtypes, sv)
-        stateordonet = widenconst(stateordonet)
     end
+    valtype = widenconst(valtype)
+    statetype = widenconst(statetype)
+    stateordonet = widenconst(stateordonet)
     if stateordonet === Nothing
         return ret
     end
```

Comment:
```suggestion
    stateordonet = abstract_call_known(interp, iteratef, nothing, Any[Const(iteratef), itertype, statetype], vtypes, sv)
    stateordonet = widenconst(stateordonet)
```
Recompute `stateordonet` with widened state.